feat: Captcha

i53577 opened the pull request:
Original PR: https://github.com/SAP/spartacus/pull/16464

i53577 commented on feature-libs/user/profile/components/register/register.component.html:
```diff
@@ -211,6 +211,10 @@
               </label>
             </div>
           </div>
+          <cx-captcha (confirmed)="captchaConfirmed()"></cx-captcha>
```
The function `captchaConfirmed()` is used to set `registerForm.captcha`. It will help to handle `cx-form-errors`.

i53577 commented on feature-libs/user/profile/occ/adapters/occ-user-profile.adapter.ts:
```diff
@@ -191,4 +199,20 @@ export class OccUserProfileAdapter implements UserProfileAdapter {
+        return config.enabled;
+      });
+
+      if (provider?.getToken() && isCaptchaEnabled) {
```
only append token when captcha is enabled.

i53577 commented on projects/storefrontapp-e2e-cypress/cypress/e2e/regression/user_access/accessibility/captcha-tabbing-order.cy.ts:
```diff
@@ -0,0 +1,33 @@
+          req.continue((res) => {
+            res?.body?.baseSites?.forEach((baseSite) => {
+              baseSite.captchaConfig = {
+                enabled: true,
```
set captcha enabled as `true` for testing

i53577 commented on projects/storefrontapp-e2e-cypress/cypress/e2e/regression/user_access/recaptchaV2.core-e2e.cy.ts:
```diff
@@ -0,0 +1,45 @@
+        req.continue((res) => {
+          res?.body?.baseSites?.forEach((baseSite) => {
+            baseSite.captchaConfig = {
+              enabled: true,
```
set captcha `enabled` as true for testing

i53577 commented on projects/storefrontapp-e2e-cypress/cypress/e2e/regression/user_access/recaptchaV2.core-e2e.cy.ts:
```diff
@@ -0,0 +1,45 @@
+      cy.wait('@getIFrame').its('response.statusCode').should('eq', 200);
+      cy.get('cx-captcha').find('iframe').should('exist');
+      registerWithCaptcha(user);
+      verifyGlobalMessageAfterRegistration();
```
Checking if a customer can be registered successfully with captcha enabled.

Zeyber commented on projects/storefrontapp-e2e-cypress/cypress/helpers/auth-forms.ts:
```diff
@@ -63,6 +63,22 @@ export function register(
+      const body = recaptchaIframe.contents();
+      cy.wrap(body).find('#recaptcha-anchor').click();
+    });
+  cy.wait(3000);
```
It is antipattern to wait for an arbiatrary amount of time. We should rather wait for something that indicates something is completed (eg. an http request/an element existing).

i53577 replied:
I used to use `cy.waitUtil()` for this condition. But seems SPA's cypress version doesn't support it. I will find another way to do this.

i53577 replied:
Fixed. Please check the latest code.

Zeyber commented on projects/core/src/model/misc.model.ts:
```diff
@@ -128,4 +128,10 @@ export interface BaseSite {
+}
+
+export interface CaptchaConfig {
+  enabled: boolean;
```
Does it make sense `enabled` optional and assume truthy or falsy if it does not exist?

i53577 replied:
According to the logic of backend commerce, it is either true or false, and it will not be null.

Zeyber commented on projects/storefrontapp-e2e-cypress/cypress.config.ci.ts:
```diff
@@ -29,6 +29,7 @@ export default defineConfig({
     OCC_PREFIX: '/occ/v2',
     OCC_PREFIX_USER_ENDPOINT: 'users',
     OCC_PREFIX_ORDER_ENDPOINT: 'orders',
+    RECAPTCHA_PUBLIC_KEY: '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI',
```
For security reasons, I would use a different key exclusive stored as a secret to the CI's domain(s). This is so that not anybody with `localhost` can use our key.

Zeyber commented on projects/storefrontapp-e2e-cypress/cypress.config.ts:
```diff
@@ -23,6 +23,7 @@ export default defineConfig({
     OCC_PREFIX: '/occ/v2',
     OCC_PREFIX_USER_ENDPOINT: 'users',
     OCC_PREFIX_ORDER_ENDPOINT: 'orders',
+    RECAPTCHA_PUBLIC_KEY: '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI',
```
Is having this key in code approved by the security team? I'm not sure another easy way to go about it in local testing but I'd like to cross off any concern they may have.

i53577 replied:
Just confirm with security guy, the answer is not. But we do need it to enable captcha for e2e test. 
Do you have any other ideas?  Can we put the publicKey in pipeline env, and get the value from it?

Zeyber replied:
Yes, it should be possible to add a secret key in the ci workflow (our github actions) using github secrets. I believe you will need to speak to an admin to do this.

Zeyber commented on feature-libs/user/profile/components/register/register.component.ts:
```diff
@@ -65,6 +65,7 @@ export class RegisterComponent implements OnInit, OnDestroy {
         this.registerComponentService.generateAdditionalConsentsFormControl?.() ??
         this.fb.array([]),
       termsandconditions: [false, Validators.requiredTrue],
+      captcha: [false, Validators.requiredTrue],
```
This is a breaking change. No user will be able to register anymore without enabling captcha. So we need to make this validator opt-in (eg. with feature flags or checking whether some captcha service has been included) until the next major version `7.0`.

i53577 replied:
Thx for your info. I will do it after confirming the release version with PO.

Zeyber commented on projects/storefrontapp-e2e-cypress/cypress/e2e/regression/user_access/recaptchaV2.core-e2e.cy.ts:
```diff
@@ -0,0 +1,45 @@
+/*
```
It could be nice to stress test our captcha feature in e2e tests so that we can make sure captcha will not fail if we do things like (for example, but does not necessarily need to be these scenarios):
- Register multiple users
- Navigate to and from registration and try register
- Register incorrectly before correctly
In essence, we want to make sure that we are not reinitializing our script or doing anything funny if we were to trigger our captcha service multiple times.

i53577 replied:
> register multiple users

why do captcha need this? The original test for register component only register one user. Is it for testing trigger captcha service multiple times?
> Navigate to and from registration and try register

this scenario has been covered. 
> Register incorrectly before correctly

Added.  Please see: https://github.com/SAP/spartacus/pull/17609/files#diff-f84f08fb6911c235a18ff25f1b7d7c2b933ab4b358ad40f741e85e151b4ec47eR75

Zeyber replied:
`Is it for testing trigger captcha service multiple times?`

Yes, exactly. I have personally had some funny behaviour from google captcha when validation is retriggered. We want to ensure this doesn't happen for us.

Zeyber commented on projects/storefrontlib/shared/components/captcha/captcha.component.ts:
```diff
@@ -0,0 +1,91 @@
+  protected subscription = new Subscription();
+
+  constructor(
+    protected config: GoogleRecaptchaApiConfig,
```
We want the captcha service to be agnostic of which captcha we are using. It should not be tied to Google's Recaptcha service. 

Please see our `TagManagementService` feature for a great example of how this can be done. In that example, the TmsService itself is agnostic but can be used for `Adobe` or `Google`'s Tag Managers interchangeably as we would like to do here.

i53577 replied:
IMHO, it doesn't make much sense to do so. SPA needs to call backend to complete the captcha feature. And the backend commerce has indicated that it uses Google's Recaptcha service. Without google recaptcha, the register process can not be completed successfully. See TAM Diagram in the following link: https://wiki.one.int.sap/wiki/display/SAPCCWiki/%5BTM%5D+Spartacus+Feature+Registration+Captcha

Zeyber replied:
While our goal has been to reach parity with the old accelerator, this is only a minimum scope. Backend has been designed with this in mind but also with the idea to expand functionality in the future. Google reCaptcha v2 is only the minimum but clients are likely to want to use captchas by different providers (even google recaptcha v3 is technically different).

We want captchas to have the possibility to be interchangable for our users and we need to develop with this in mind (Spartacus is all about extensibility afterall). We only want to use this Google captcha as a default for Spartacus, but also leave the possibility for it to be swapped our for a different configuration. Therefore, I strongly believe we should decouple `captcha` from `google-captcha-v2` as much as possible.

Does it make sense?

Zeyber reviewed (changes requested):
One more e2e test that could be good to make sure Captcha is working correctly (covering all our basic):
- Is it possible to test that Captcha will block registration if the public key is incorrect?

kpawelczak committed:
chore: update node version (#18455)

kpawelczak committed:
feat: create migration mechanism - update ssr (#18430)

Automates 2211.19 ssr migration after angular 17 migration schematics changed it:
- removes server.ts and renames server.ts.bak
- removes express.tokens
- fixes imports in server.ts and any import related to REQUEST, RESPONSE

kpawelczak committed:
docs: add migration step to 2211-migration.md (#18459)

anjana-bl committed:
feature/CXINT-2714: Show promotions on PDP page (#18456)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

kpawelczak committed:
chore: update migration.json with 2211.19 scripts (#18460)

Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>

kpawelczak committed:
Generate breaking changes migrations (#18442)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>

pawelfras committed:
chore: CXSPA-6086 Code cleanup - ticket references (#18469)

kpawelczak committed:
docs: update migration.md with devkit/schematics installation (#18472)

kpawelczak committed:
chore: remove isLevel flag in ts files (#18468)

pawelfras committed:
doc: CXSPA-5921 Remove HttpClientModule from ClientAuthStoreModule - migration docs (#18457)

Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>

kpawelczak committed:
feat: install spartacus/schematics dev dependencies during spa installation (#18473)

Platonn committed:
chore: remove cxFeatureLevel usages (#18466)

Co-authored-by: Paweł Fraś <fras.pawel@yahoo.com>
Co-authored-by: kpawelczak <42094017+kpawelczak@users.noreply.github.com>
Co-authored-by: Hak Woo Kim <kimhakwoo@hotmail.com>

kimhw0630 committed:
doc: documentation change for ASM library (CXSPA-6017) (#18449)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Florent Letendre <florent.letendre01@sap.com>

kimhw0630 committed:
doc: document for updates on type incongruities (CXSPA-6022) (#18465)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

kpawelczak committed:
chore: run breaking changes generator (#18474)

kimhw0630 committed:
doc: SSR Optimization cacheSize update (CXSPA-6023) (#18467)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

kimhw0630 committed:
doc: document for ASM CustomerSelectionComponent (CXSPA-6018) (#18475)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

kimhw0630 committed:
doc: document for address book (CXSPA-6020) (#18477)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

kpawelczak committed:
fix consignmentTracking feature flag (#18478)

i53577 committed:
feat: my account v2 

My Account v2 views including:
- My Information
- My Email Address
- Update Password and Security
- Update Communications (Preferred Communication Channel)
- Privacy And Settings (Consent Management)

kpawelczak committed:
fix: add browser-sync (#18480)

kpawelczak committed:
fix: add browser-sync (#18481)

rmch91 committed:
chore: Script for testing all snaps (#18482)

kpawelczak committed:
docs: add a mention of removed $useLatestsStyles from pick-up-in-store  (#18484)

kimhw0630 committed:
doc: Remove feature level for requested delivery date (CXSPA-6028) (#18485)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

lasoh committed:
feat: use recent searches (#18491)

Add new functionality to search box - recent searches

anjana-bl committed:
feature/CXINT-2740: Disable editing when service requests are from CDP (#18470)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

pawelfras committed:
doc: CXSPA-5811 SSR contextual logging - changes after making this feature default (#18486)

Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>

pawelfras committed:
chore: CXSPA-6159 Adjust migration steps in 2211-migration.md file (#18503)

rmch91 committed:
Release 2211.19.0 (#18504)

anjana-bl committed:
test: re-naming the e2e file to exclude it from pipeline (#18508)

pawelfras committed:
feat: quote support (#16345)

Co-authored-by: Larisa Staroverova <larisa.staroverova@sap.com>
Co-authored-by: Ulrich Stellmacher <ulrich.stellmacher@sap.com>
Co-authored-by: Christoph Hinssen <christoph.hinssen@sap.com>
Co-authored-by: steinsebastian <54309425+steinsebastian@users.noreply.github.com>
Co-authored-by: Moritz Schäfer <Moritz.schaefer@sap.com>

ChristophHi committed:
chore: define code owners for quote feature lib (#18509)

i53577 committed:
fix: my account v2 e2e tests failure (#18511)

gladius-mtl committed:
doc: update readme and contributing docs (#18515)

i53577 committed:
fix: add traceability for my account v2 e2e tests (#18516)

ChristophHi committed:
fix: ensure OV form takes ghost styling changes into account (#18517)

rmch91 committed:
Merge remote-tracking branch 'origin/develop-next-major' into develop

rmch91 committed:
chore: Update config (#18524)

niehuayang committed:
fix: change the style for Agent login page (#18518)

i53577 committed:
fix: remove unused logic for my account v2 e2e tests (#18519)

rmch91 committed:
Release 2211.20.0-1 (#18528)

niehuayang committed:
fix device compatibility Issues for My Account v2 (#18529)

ChristophHi committed:
chore: add option to install quote feature lib (#18536)

rmch91 committed:
fix: Increase maximumError to 3.5mb (#18540)

pawelfras committed:
fix: CXSPA-6327 Use ExpressServerLogger type when logger is initialized in OptimizedSsrEngine (#18537)

rmch91 committed:
Release 2211.20.0 2 (#18542)

ChristophHi committed:
fix: add cxAbstractOrderContext in 2 places where missing (#18544)

alepil committed:
chore: fix failing ccv2 cpq e2e test on p3 (18538) (#18538)

This fix solves https://jira.tools.sap/browse/CXSPA-4979
The cypress test waits for the wrong the wrong value and fails.
Additional hardcoded urls are removed from the calling ccv2-wtf script as documented in https://jira.tools.sap/browse/CXEC-36133

Closes #18538

kimhw0630 committed:
fix: navigation issue in the paragraph component (#18533)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Paweł Fraś <fras.pawel@yahoo.com>

kimhw0630 committed:
fix: Title field is not mentioned as an optional (CXSPA-5943) (#18543)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

alepil committed:
chore: fix product-configuration b2c/b2b/cpq test scans on ccv2 environment (18550) (#18550)

fixes failing scans like https://gkeeccommercecloud.jaas-gcp.cloud.sap.corp/job/testing-framework/job/test-runner/49401/display/redirect

Closes #18550

ChristophHi committed:
chore: move styling change to quote (to avoid feature toggling) (#18549)

dependabot[bot] committed:
chore(deps): bump undici and @angular-devkit/build-angular (#18552)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>

dependabot[bot] committed:
chore(deps): bump ip from 2.0.0 to 2.0.1 (#18551)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

sam-garland committed:
fix(CXSPA-4793): replace url with obviously fake one (#18553)

steinsebastian committed:
fix: Accessibility fixes for tab-bar and add-to-cart button in configuration (#18530)

giancorderoortiz committed:
Scan develop branch for OSS lib updates (#18561)

Stofftiger committed:
test: fix failing cypress tests  (#18558)

renovate[bot] committed:
chore(deps): update styfle/cancel-workflow-action action to v0.12.1 (#18565)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

Uli-Tiger committed:
fix: cpq e2e tests (CXSPA-5302) (#18570)

anjana-bl committed:
fix/CXSPA-6380: My Account V2 Order History does not load when details API throw 400 status code (#18562)

steinsebastian committed:
fix: Enable enter/space-key for parent-/sub-group navigation for better accessibility (#18560)

Co-authored-by: Larisa Staroverova <larisa.staroverova@sap.com>

lasoh committed:
fix: recent searches cxfeature flag  (#18578)

Co-authored-by: lasoh <artur.lasocha@pekao-fs.com.pl>

ChristophHi committed:
fix: set proper changeDetection for handling hostBinding changes (#18580)

ChristophHi committed:
fix: avoid not needed OCC calls  calls on configurator refocus  (#18601) (cxspa-6450)

Co-authored-by: Sebastian Stein <sebastian.stein@sap.com>

giancorderoortiz committed:
Express OSS Lib Bump from 4.18.2 to 4.18.3 (#18610)

pawelfras committed:
bugfix: CXSPA-6445 OCC_BACKEND_BASE_URL_VALUE not mapped to value in new SPA 2211.19 app with SSR (#18602)

pawelfras committed:
fix: CXSPA-6456 Missing  script in SSR app causes errors during the CCv2 build (#18604)

kpawelczak committed:
feat: change breaking changes script so it would work with new spa versions (#18496)

kpawelczak committed:
feat: update defaultRenderingStrategyResolver to disable ssr in smartedit (#18559)

Platonn committed:
feat: named feature flags mechanism for css styling (#18523)

For feature flags in styles we will no longer use mixin `@include forVersion('x.y')`, but `@include forFeature('stringFeatureName')`, where the string feature name is synchronized with Spartacus global config `provideConfig({ features: { ... } })`.

Architecture overview:
- `FeatureStylesService` helper service in TS - applies/removes feature flags CSS classes on the root element, but only  _on demand_, when specific feature flags are _actually used_ on the current page
- `useFeatureStyles(feature)` function in TS  - called directly in components' constructors; it demands setting the relevant feature flag CSS class in the root element. (note: Unsetting the class is done automatically on components' destroy). It uses helper service `FeatureStylesService`, which under the hood counts usage references of each flag. 
- `forFeature(feature){ ... }` mixin in SCSS - adds extra CSS styles, for a selector `.cxFeat_{feature} &` - in other words: when `cxFeat_{feature}` is set on the ancestor element (implicitly: on the root element) then extra CSS styles will apply for the current selector `&`

Notes:
- feature flag CSS class shape: `cxFeat_{feature}`.
- the root element is Angular's root element, but not body. Philosophical concern: Angular app should not own body, but only it's own root element (i.e. cx-storefront)
- using mixin `forFeature()` directly inside styles of the root component (i.e. cx-storefront) won't work. In this speciic edge case, you have to add a special 2nd string param for the mixin: `'currentSelector'`, for example:
  ```scss
  %cx-storefront {
    @include forFeature('darkMode', 'currentSelector') {
      filter: invert(80%);
    }
  }
  ```

**Pros:**
- feature flag CSS is set on the root element **only if it's used anywhere** (only if any currently rendered component declares `useFeatureStyles(feature)`. Thanks to it, we avoid for example setting on on homepage any CSS feature flags related only to /my-account features.

**Cons:**
- developers needs to remember to declare `useFeatureStyles(feature)` in the component's constructor, if they want to use feature's styles
- [edge case] when using `forFeature()` SCSS mixin in the root element's styles, the 2nd param of string value `'currentSelector'` needs to be passed to the SCSS mixin `forFeature(..., 'currentSelector')`

related to https://jira.tools.sap/browse/CXSPA-6042

scarai-sap committed:
CXSPA-6263: Fix issue of styling (#18619)

rmch91 committed:
Release 2211.20.0 (#18621)

rmch91 committed:
fix: CXSPA-6438 Add isPlatformBrowser check before OAuth initialization  (#18608)

kimhw0630 committed:
refactor: remove using private property (CXSPA-6439) (#18605)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

renovate[bot] committed:
chore(deps): update dependency @types/semver to v7.5.8 (#18563)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

Larisa-Staroverova committed:
feat: support read-only attribute types with images (#18415) (CXSPA-5910)

Larisa-Staroverova committed:
feat: display description at configurator attribute and value level  (#18245) (CXSPA-3242)

Co-authored-by: Christoph Hinssen <christoph.hinssen@sap.com>

Zeyber committed:
fix: Fix unit addresses and users in b2b my company not traversing or notifying on creation (#18493)

Closes: https://jira.tools.sap/browse/CXSPA-5630

kimhw0630 committed:
chore: update product scope for product carousel(CXSPA-4721) (#18598)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

Akash-SAP committed:
CXINT-1000: Toast Message to show profile update (#18589)

rmch91 committed:
chore: Workflow for minor release (#18594)

Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency @angular-eslint/template-parser to v17.2.1 (#18567)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

kpawelczak committed:
feat: replace major mentions in breaking changes script (#18632)

Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>

kimhw0630 committed:
chore : add/update eslint rule (CXSPA-5988) (#18568)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

rmch91 committed:
chore: Restore previous version of repo-sync (#18631)

rmch91 committed:
fix: Use version env variable in publishing version const (#18637)

ChristophHi committed:
fix: remove authGuard from 2 quote components (#18633)

Pio-Bar committed:
feat: (CXSPA-1026) - Added asterisks to required form fields (#18264)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: Caine Rotherham <rotherham.c@gmail.com>

Pio-Bar committed:
fix: (CXSPA-954) - Restored default tab flow for quantity counter (#18448)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-2189) - Order confirmation heading order (#18531)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

kpawelczak committed:
Release 2211.21.0-1 (#18638)

Pio-Bar committed:
fix: (CXSPA-924) - Language dropdown aria-label (#18441)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: Moved featureConfig import to the correct quick order module (#18642)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1027) - Improve a11y of star rating component (#18268)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1770) - Improved aria attributes for table buttons (#18479)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-965) - Refocus on input after quick order with keyboard (#18313)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Larisa-Staroverova committed:
fix: rename product configurator feature toggle

Larisa-Staroverova committed:
test: fix Core & Mock e2e tests (#18651)

Closes CXSPA-6589

ChristophHi committed:
fix: reset loading state if variant search feature is not active (#18654)

steinsebastian committed:
fix: Focus handling in group menu in mobile mode (#18647)

Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>

Uli-Tiger committed:
fix: Overview page not loading for CPQ Configurator [CXSPA-6590] (#18656)

Pio-Bar committed:
feat: (CXSPA-981) - Add keyboard navigation to 'navigation-ui' (#18297)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

Pio-Bar committed:
fix: (CXSPA-2245) - Aria-live for org-message (#18617)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: Enabled a11y feature flags (#18661)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-955) - Restore focus after delivery modes are refetched (#18444)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

Pio-Bar committed:
fix: Focus management e2e fix (#18671)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1023) - Changed button text to be consistent with modal title (#18653)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1021) - Login screen placeholder translations change (#18668)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

KateChuen committed:
fixing password-visibility e2e test (#18673)

Pio-Bar committed:
fix: (CXSPA-1047) - stop focus indicator from expanding (#18277)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1051) - Stop sorting dropdown from truncating options (#18423)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pawulon-0 committed:
fix: Make missing actions column header visible in saved cart list (#18426)

Co-authored-by: Paweł Kwit <pawel.kwit@agcocorp.com>
Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: PioBar <72926984+Pio-Bar@users.noreply.github.com>

Pawulon-0 committed:
fix: (CXSPA-1049) - Keep focus on scroll to top button when clicked by keyboard (#18532)

Co-authored-by: Paweł Kwit <pawel.kwit@agcocorp.com>
Co-authored-by: rmch91 <r.chygryn@sap.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: PioBar <72926984+Pio-Bar@users.noreply.github.com>

Pio-Bar committed:
fix: (CXSPA-2187) - Schedule replenishment order a11y fixes (#18525)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-2248) - Checkout schedule replenishment order Fieldset (#18596)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1794) - List heading order (#18464)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-2243) - Improved a11y for popovers in 'organization' (#18539)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Larisa-Staroverova committed:
fix: switch default for storeFrontLibCardParagraphTruncated feature toggle (#18674)

Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>

Pio-Bar committed:
fix: (CXSPA-2684) - Fixed oversized focus ring for table cells (#18603)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1057) - Cart summary heading order (#18439)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

sdrozdsap committed:
fix: (CXSPA-1055) - Indication of required checkbox in checkout (#18652)

kimhw0630 committed:
fix: fix sonar issues (#18679)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

Pio-Bar committed:
fix: (CXSPA-2186) - Aria label input for generic link (#18520)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1031) - Map overflow on small/zoomed screens (#18391)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: Updated e2e selectors to match recent DOM changes (#18684)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

dependabot[bot] committed:
chore(deps): bump express from 4.18.3 to 4.19.2 (#18692)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>

dependabot[bot] committed:
chore(deps-dev): bump webpack-dev-middleware from 5.3.3 to 5.3.4 (#18694)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>

kpawelczak committed:
fix: Changing language in saved cart page now reloads saved carts (#18685)

After change we are loading cart data once again on site context change.

Closes: https://jira.tools.sap/browse/CXSPA-6551

dependabot[bot] committed:
chore(deps-dev): bump follow-redirects from 1.15.5 to 1.15.6 (#18693)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

sdrozdsap committed:
fix: (CXSPA-3368) - misplaced aria hidden attrs on focusable elements (#18634)

giancorderoortiz committed:
Sec. Update angular-devkit/build-angular to v17.3.2 (#18699)

Larisa-Staroverova committed:
fix: Handle read-only correctly (#18690)

Closes CXSPA-6663

anjana-bl committed:
Bug/CXINT-2438: Update Sort Keys in Spartacus PDF Invoices Component  (#18488)

RadhepS committed:
Release 2211.21.0 (#18703)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

Platonn committed:
fix: schematics replace local AppRoutingModule with Spartacus' AppRoutingModule (#18708)

**BEFORE:**

Spartacus installation schematics silently failed to add the Spartacus' `AppRoutingModule` import to the customer's app.module. Instead the Angular's generated `AppRoutingModule` (yes, the same name) remained in the app.module. This caused the Angular's `RouterModule` to be misconfigured - it didn't meet the requirements of Spartacus (e.g. missing options: `RouterModule.forRoot([], { initialNavigation, anchorScrolling} )`).

**NOW:**

Now we added a fix: we replace the local file `AppRoutingModule` (created by Angular by default) with our Spartacus `AppRoutingModule`

**WHY:**

We noticed it only now. But since ng17 Angular creates new applications with "routing" option enabled by default (see ng17 commit: https://github.com/angular/angular-cli/commit/1a6a139aaf8d5a6947b399bbbd48bbfd9e52372c). Because of that, new Angular 17 applications contain a local file `AppRoutingModule` which clashes (by name) with our schematics' attempt to add our `AppRoutingModule` **from Spartacus**.

The problem did not occur before Angular v17, because apps created didn't use to contain a local `AppRoutingModule` (due to "routing" option being disabled by default in the `ng new` command).


Note: now in all unit tests of our schematics we use an Angular app with implicit default "--routing=true", to follow the defaults of Angular in our default test cases

Note 2: to cover the opposite case too, we added now a special unit test for our schematics, where we use an Angular app with deliberate option "--routing=false"  too

**QA steps:**
**1. With routing implicitly enabled**

- build libs locally
- publish libs to verdaccio, e.g. using `schematics/testing.ts` script
- `npx @angular/cli@17 new test-ng17-spa-AppRoutingModule --standalone=false` (will install with default routing:true)
- verify in `app.module` you have `import { AppRoutingModule } from './app-routing.module';`
- `cd test-ng17-spa-AppRoutingModule`
- `ng add @spartacus/schematics`
- verify in `app.module` you have `import { AppRoutingModule } from '@spartacus/storefront';`
- verify in `app.module` the `imports` array of `AppModule` contains `AppRoutingModule`

Diff of `app.module` focused on `AppRoutingModule` import:
![image](https://github.com/SAP/spartacus/assets/4001059/dcff926d-fd7f-40eb-a732-e69919ffccf4)

Diff of the whole `app.module` (just for reference):
![image](https://github.com/SAP/spartacus/assets/4001059/e90b9cef-ee2f-44d8-a844-762ba95837fd)

**2. With routing explicitly disabled**

- Repeat all steps from point (1), but pass flag `--no-routing` to the command `npx @angular/cli@17 new ....`

Diff of `app.module` focused on `AppRoutingModule` import:
![image](https://github.com/SAP/spartacus/assets/4001059/65f4f54f-a1cf-4971-a191-756e9b714d02)

Diff of the whole `app.module` (just for reference):
![image](https://github.com/SAP/spartacus/assets/4001059/904eda2a-dd89-4fd2-bea7-769f14fae62e)


closes https://jira.tools.sap/browse/CXSPA-6726

pawelfras committed:
fix: CXSPA-6581 Race condition in OpenID code flow causes redirect loop or double redirect on login (#18706)

Pio-Bar committed:
fix: (CXSPA-2338) - Reorder dialog a11y fixes (#18648)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-2798) - Show assistive message when active view changes (#18609)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

rmch91 committed:
chore: Change release branch pattern in CI (#18698)

sdrozdsap committed:
fix: (CXSPA-1178) - Display/pronounce password requirements (#18711)

Pio-Bar committed:
fix: (CXSPA-1030) - Visible focus fixed on mobile/while zoomed (#18382)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

pawelfras committed:
feat: CXSPA-6765 Add LOCATION_INITIALIZED_MULTI to public API (#18726)

RadhepS committed:
chore: update scripts used for e2e tests on release pipeline 

Add scripts to allow for release e2es to run differently

closes: https://jira.tools.sap/browse/CXSPA-5290

Platonn committed:
feat: new `FeatureToggles` config to drive parts of Spartacus config, e.g. cms mapping (#18670)

The new `FeatureToggles` config is separate from Spartacus config, to be able to control the global Spartacus config. 
The `FeatureToggles` is a simple list of boolean properties which allows for enabling breaking changes in Spartacus, ALSO changes in the default config chunks like cms mapping etc. See example:
```ts
provideDefaultFeatureToggles({
  // boolean is a simple, serializable flag, so can be even potentially
  // fetched from remote configuration endpoint in runtime
  myAccountV2: true
})
```
```ts
// conditionally changing the default config chunk based on the simple boolean feature toggle:
provideConfigFactory(()=>{
  const featureToggles = inject(FeatureToggles);
  return featureToggles.myAccountV2
    ? { cmsComponents: { component: MyAccountComponentV2 } }
    : { cmsComponents: { component: MyAccountComponent } };
})
```

In other words, the `FeatureToggles` config can be a bridge between a simple (and even remote!) feature toggles configuration of Spartacu and making actual changes into the default config chunks in Spartacus (like default cms mapping).

Note: In this PR added also an ESLint rule to prefer using in our libraries the function `provideDefaultFeatureToggles()` over `provideFeatureToggles()` - so our providers have lower precedence than customers' ones.

**ALL FEATURE TOGGLES DEFINED IN THE CORE LIB:**

All new toggles are now added in the file `feature-toggles.ts` for customers and in `spartacus-features.module.ts` for our storefront app.

The centralization of feature toggles was done because:
- for doc writers: otherwise It was hard to track down new toggles each release due to distribution across many libs
- for devs: adding a new toggle lacked a standard process, raising the question: Where to define it? This dilemma leads to issues, **especially when default toggles values were provided by mistake in lazy-loaded modules it was not working** (because the feature toggles mechanism relies on the **root injector**).
- this confusion wasted developers' time, as they "just want to add a toggle" but get stuck figuring out where to place it, with the process varying across different libraries (storefront, core, feature).


**DEPRECATIONS:**

This PR deprecates the `FeaturesConfig` interface in favor of `FeatureToggles`: So please use:
```ts
provideFeatureToggles({
  [string]: boolean
})
```
instead of
```ts
provideConfig({
  features: {
    [string]: boolean
  }
})
```
That said, the old approach will still work until we remove the `FeaturesConfig` eventually.

**QA steps**

I) verify that FeatureToggles mechanism work as expected on the example property `a11yRequiredAsterisks`:
1. open login page
2. verify that the asterisk regarding required form fields IS displayed
3. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify asterisk is NOT displayed
4. in `default-feature-toggles.ts` set `a11yRequiredAsterisks: true` and verify asterisk IS displayed
5. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: false` and verify asterisk is NOT displayed
6. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: true` and verify asterisk is IS displayed
7. in `default-feature-toggles.ts` set `a11yRequiredAsterisks: false` and verify asterisk IS STILL displayed
8. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify asterisk is NOT displayed

II) verify FeatureToggles mechanism can alter a cms mapping, let's use `a11yRequiredAsterisks` for the spike purposes:
1. in `login-form.module.ts` add to the `providers` array the following code:
```ts
    provideDefaultConfigFactory(() => {
      return inject(FeatureToggles).a11yRequiredAsterisks
        ? {
            cmsComponents: {
              ReturningCustomerLoginComponent: {
                component: SpikeComponent,
              },
            },
          }
        : {};
    }),
```
2. ...and in the top of the same file (`login-form.module.ts`) add the following code:
```ts
@Component({
  selector: 'cx-spike',
  template: 'SPIKE COMPONENT WITH ASTERISK (*.*)',
})
export class SpikeComponent {}
```

3. Visit the login page and verify that the `SPIKE` text is displayed
4. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify that the `SPIKE COMPONENT WITH ASTERISK (*.*)` text is NOT displayed

III) verify the old FeaturesConfig mechanism cannot alter cms mapping, but still works for enabling features:
1. ensure that `a11yRequiredAsterisks` in `provideFeatureToggles` is set to `false`
2. in `spartacus-features.module.ts` add to the `providers` array the following code `provideConfig({features: { a11yRequiredAsterisks: true  }})` and verify that the ORIGINAL LOGIN COMPONENT is displayed WITH real a11y asterisk feature displayed
3. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: false` inside the above mentioned `provideConfig()` and verify that the ORIGINAL LOGIN COMPONENT is displayed WITHOUT real a11y asterisk feature displayed
4. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: true` inside the `provideFeatureToggles()` and verify that the the `SPIKE` text is displayed even if `a11yRequiredAsterisks: false` inside the `provideConfig()` is false


closes https://jira.tools.sap/browse/CXSPA-6586

kimhw0630 committed:
remove Unused variable (#18701)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

ChristophHi committed:
chore: fix storefrontapp value for toggle productConfiguratorAttributeTypesV2 (#18735)

kimhw0630 committed:
release 2211.23.0-1 (#18739)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

sdrozdsap committed:
fix: (CXSPA-1017) - Phone number placeholder styles (#18729)

sdrozdsap committed:
fix: (CXSPA-1054) - Provide more meaningful placeholder for apt number (#18728)

sdrozdsap committed:
fix: (CXSPA-1029) - Image text alternative is provided to banners (#18719)

sdrozdsap committed:
fix: (CXSPA-940) - Display tooltips for non text content (#18733)

Pio-Bar committed:
doc: JSDocs for a11y feature toggles (#18737)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Larisa-Staroverova committed:
docs: JSDocs for product configurator feature toggle

RadhepS committed:
fix: update major version in SCSS (#18755)

fix: update major version in SCSS

closes: https://jira.tools.sap/browse/CXSPA-6735

Zeyber committed:
fix: Fix AppRouting configurations not being active when using install script (#18717)

Closes: https://jira.tools.sap/browse/CXSPA-6158

sdrozdsap committed:
fix: (CXSPA-1104) - Provide more meaningful text for order status (#18751)

Pio-Bar committed:
fix: (CXSPA-4344) - Quantity update on enter press (#18713)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Platonn committed:
fix: enable all existing Feature Toggles in new apps via installation schematics (#18744)

The installation schematics now generates a list of available FeatureToggles in a customer's app.
It's to help customers to see what feature toggles are available in Spartacus.

Moreover, each feature toggle is explicitly enabled in customer's app, to help customers start with the most up-to-date configuration (without a technical debt of any feature toggles disabled). Of course, they can disable any feature toggle they don't want to use.

In particular, our installation schematics generates the list of FeatureToggles right in the `SpartacusFeaturesModule` module, in the `providers` array, in the client's app - all in one place.

For new apps, it's not a breaking change because they're creating a new app, not upgrading an existing one.

Note: we're not enabling any new feature toggles for existing apps. This PR is only about the installation schematics for new apps.

**QA steps**
1. build libs and publish to verdaccio
2. create new app `npx @angular/cli@17 new test-ng17-FeatureToggles
 --standalone=false`
3. add Spartacus  `ng add @spartacus/schematics --baseUrl="https://40.76.109.9:9002"`
4. verify the bottom of the `spartacus-features.module.ts` in the new app contains the enablement of all feature toggles properties (compare it with the the object `defaultFeatureToggles` from Spartacus repo):
```typescript
providers: [
  provideFeatureToggles:({
     <JSON object with all feature toggles here; each with value true>
  })
]
```

fixes https://jira.tools.sap/browse/CXSPA-6783

dependabot[bot] committed:
chore(deps): bump tar from 6.2.0 to 6.2.1 (#18718)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump vite and @angular-devkit/build-angular (#18762)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

LindaSAP committed:
CXSPA-6906: Make getting login form exist after alert message show in E2E myaccountv2 email test (#18766)

Platonn committed:
chore: fix build/test of schematics in individual libraries, when running it in isolation from the core schematics build/test (#18769)

Previously we're run the script "copy-feature-toggles" only as a pre-build (and a pre-test) step only of the core schematics. 

Now we run script "copy-feature-toggles" inside the script "clean" of the core schematics.

Why: All individual libraries' "test:schematics" commands run the core's schematics "clean" script before executing their tests. So now we're sure that the script "copy-feature-toggles" will be run before "test:schematics" of any individual library. Such an individual run of tests might happen due to `Nx` choosing to build only the selective changed libraries.

**QA steps:**
Verify you can run tests of schematics of an individual lib, e.g. QuotesL
1. checkout to this branch locally
2. make sure you DO NOT HAVE a file `projects/schematics/src/feature-toggles.copied-from-core-lib.ts` - otherwise please delete it.
3. `cd feature-libs/quote`
4. `npm run test:schematics`
5. Verify the tests passed. Especially no such error occured like there: https://github.com/SAP/spartacus/actions/runs/8815031194/job/24196138821?pr=18768

ChristophHi committed:
fix: browser back shows correct quote details after 'new cart' (#18768)

Pio-Bar committed:
fix: (CXSPA-6770) - Aria-label for popover close button (#18760)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

sdrozdsap committed:
fix: (CXSPA-939) - Buttons should be triggered by space and enter (#18740)

sdrozdsap committed:
fix: (CXSPA-1109) - Styles of links for cart table items (#18750)

Pio-Bar committed:
feature: (CXSPA-2190) - Keyboard controls for organization tables (#18514)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

rmch91 committed:
fix: Fix image lazy loading for Chrome in SSR mode (#18761)

This PR provides the work around for not working image lazy loading in SSR in Chrome and Safari.

Closes: https://jira.tools.sap/browse/CXSPA-6719

sdrozdsap committed:
test: Failing scroll-position-restoration spec (#18772)

kimhw0630 committed:
fix: tag event shouldn't trigger if consent is unavailable (CXSPA-6932) (#18771)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>

Reid-McNaughton committed:
fix: update epd-visualization e2e test to not rely on tabindex state of footer controls (#18764)

Co-authored-by: Sam Garland <52560012+sam-garland@users.noreply.github.com>

anjana-bl committed:
BugFix/CXSPA-6505: Issue with CDC Schematics (#18702)

sdrozdsap committed:
fix: (CXSPA-1044) - Carousel controls accessible with keyboard (#18759)

kimhw0630 committed:
fix:scroll-position-restoration e2e (#18773)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

renovate[bot] committed:
chore(deps): update dependency qs to v6.12.0 (#18630)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>

KateChuen committed:
test cypress upgrade to 13.8.0 (#18784)

scarai-sap committed:
feat/CXSPA-6745: Bind Cart: move sensitive data in body (#18738)

steinsebastian committed:
fix: Fix accessibility issues in conflict solver  (#18775)

KateChuen committed:
Revert "test cypress upgrade to 13.8.0" (#18788)

dependabot[bot] committed:
chore(deps-dev): bump ejs from 3.1.9 to 3.1.10 (#18791)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>

renovate[bot] committed:
chore(deps): update dependency typescript to v5 (#18792)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency jasmine-core to v5 (#18795)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

RadhepS committed:
chore: mark express checkout e2e as flaky (#18797)

PR to fix e2e:  https://jira.tools.sap/browse/CXSPA-7064

Larisa-Staroverova committed:
feat: define cartQuickOrderRemoveListeningToFailEvent feature toggle (#18780)

Define cartQuickOrderRemoveListeningToFailEvent feature toggle to avoid showing an unnecessary duplicated error message on the failure of adding to cart

ChristophHi committed:
fix: addedToCartDialog driven by success event, takes actual product code into account (#18731)

Larisa-Staroverova committed:
feat: make-to-stock product variants (#18505)

RadhepS committed:
chore: create auth flow e2e (#18805)

chore: Create auth flow e2e

Closes: https://jira.tools.sap/browse/CXSPA-6722

steinsebastian committed:
CXSPA-6099: fix accessibility issues for read-only attributes (found … (#18807)

ChristophHi committed:
fix: addedToCartDialog driven by initiation and success event

renovate[bot] committed:
chore(deps): update actions/cache action to v4 (#18809)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

Platonn committed:
feat: create `BEFORE_CMS_PAGE_GUARD` injection token allowing for custom pre-`CmsPageGuard` logic (#18781)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: anjana-bl <anjana.b.l@sap.com>
Co-authored-by: Paweł Fraś <fras.pawel@yahoo.com>

Platonn committed:
chore: remove obsolete "resolutions" and "overrides" entries from package.json - because @nguniversal is no longer used (#18804)

fixes https://jira.tools.sap/browse/CXSPA-7101

renovate[bot] committed:
chore(deps): update dependency stringify-object to v5 (#18810)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

RadhepS committed:
chore: update dependency sass-true to v8 (#18821)

chore: update dependency sass-true to v8

closes: https://jira.tools.sap/browse/CXSPA-7129

anjana-bl committed:
feat/CXSPA-6636: OPPS Coupon Codes & Login required (#18803)

Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

kimhw0630 committed:
update version to 2211.23.0 (#18831)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>
Co-authored-by: Radhep Sabapathipillai <radhep.sabapathipillai@sap.com>

sdrozdsap committed:
fix: (CXSPA-1065) - Hide select button for selected address or payment (#18774)

tiwariakshay55 committed:
CXSPA-6816 : Estimated delivery date UI (#18794)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: LindaSAP <153066064+LindaSAP@users.noreply.github.com>
Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>
Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: PioBar <72926984+Pio-Bar@users.noreply.github.com>
Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: sdrozdsap <163305268+sdrozdsap@users.noreply.github.com>
Co-authored-by: Roman <129765378+rmch91@users.noreply.github.com>
Co-authored-by: Hak Woo Kim <kimhakwoo@hotmail.com>
Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>
Co-authored-by: Reid-McNaughton <99927305+Reid-McNaughton@users.noreply.github.com>
Co-authored-by: Sam Garland <52560012+sam-garland@users.noreply.github.com>
Co-authored-by: anjana-bl <84384970+anjana-bl@users.noreply.github.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>
Co-authored-by: KateChuen <45409515+KateChuen@users.noreply.github.com>
Co-authored-by: Scar Ai <38693320+scarai-sap@users.noreply.github.com>
Co-authored-by: steinsebastian <54309425+steinsebastian@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

lasoh committed:
feat: Keyword redirect support in Spartacus (#18667)

Co-authored-by: Artur Lasocha <artur.lasocha@sap.com>
Co-authored-by: Peter Kurajsky <pekura@users.noreply.github.com>
Co-authored-by: Peter Kurajsky <peter.kurajsky@sap.com>

RadhepS committed:
chore: Revert "CXSPA-6816 : Estimated delivery date UI" (#18845)

Reverts #18794

KateChuen committed:
fix for the language & currency checkout e2es (#18847)

Pio-Bar committed:
fix: (CXSPA-6857) - Footer tabbing fix (#18777)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

sdrozdsap committed:
fix: (CXSPA-1182) - Truncated text on mobile view (#18817)

sdrozdsap committed:
fix: (CXSPA-1109) - Product name link styles on my interests page (#18818)

Pio-Bar committed:
feat: (CXSPA-946) - Keyboard navigation for FacetComponent (#18758)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-975) - Removed unnecessary aria-live attributes (#18833)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

steinsebastian committed:
fix: aria-label not associated with button in mobile mode (#18836)

sdrozdsap committed:
fix: (CXSPA-953) - my account link outline overlapping with user name (#18825)

sdrozdsap committed:
fix: (CXSPA-950) - Close product image button focus is visible (#18832)

sdrozdsap committed:
fix: (CXSPA-1009) - Pronounce CVV input tooltip information (#18834)

kimhw0630 committed:
fix product search product e2e (#18851)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

kimhw0630 committed:
fix: update customer 360 and b2b saved cart e2es (#18854)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

i333055 committed:
feat: CXSPA-6689 one time password for login (#18752)

Co-authored-by: niehuayang <niehua.yang@sap.com>
Co-authored-by: Niehua Yang <126219214+niehuayang@users.noreply.github.com>

Larisa-Staroverova committed:
test: Fix failing CPQ cypress tests (#18857)

Closes CXSPA-7262

Zeyber committed:
test: SSR E2E Testing Framework (#18779)

Closes: CXSPA-6525

Pio-Bar committed:
fix: (CXSPA-1166) - Focus on hidden search box (#18770)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1161) - Visible focus obstruction and overflows fix. (#18830)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

kpawelczak committed:
feat: utlity function to color the text (#18853)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

kimhw0630 committed:
chore: fix e2e after upgrade cypress version to 13.8.0 (CXSPA-6957) (#18799)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

Larisa-Staroverova committed:
Fix timeout for quote cypress tests (#18861)

renovate[bot] committed:
chore(deps): update dependency webpack to ~5.91.0 (#18826)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

Uli-Tiger committed:
fix: each cy test shall use unique user  (CXSPA-7213) (#18860)

Larisa-Staroverova committed:
feat: Configure from quote cart: Prompts and messages should refer to quote (#18806)

Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: Christoph Hinssen <christoph.hinssen@sap.com>

tiwariakshay55 committed:
CXSPA-6816 : Estimated delivery date UI (#18865)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: LindaSAP <153066064+LindaSAP@users.noreply.github.com>
Co-authored-by: Krzysztof Platis <platonn.git@gmail.com>
Co-authored-by: Christoph Hinssen <33626130+ChristophHi@users.noreply.github.com>
Co-authored-by: PioBar <72926984+Pio-Bar@users.noreply.github.com>
Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>
Co-authored-by: sdrozdsap <163305268+sdrozdsap@users.noreply.github.com>
Co-authored-by: Roman <129765378+rmch91@users.noreply.github.com>
Co-authored-by: Hak Woo Kim <kimhakwoo@hotmail.com>
Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>
Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>
Co-authored-by: Reid-McNaughton <99927305+Reid-McNaughton@users.noreply.github.com>
Co-authored-by: Sam Garland <52560012+sam-garland@users.noreply.github.com>
Co-authored-by: anjana-bl <84384970+anjana-bl@users.noreply.github.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <giancarlo.cordero.ortiz@sap.com>
Co-authored-by: KateChuen <45409515+KateChuen@users.noreply.github.com>
Co-authored-by: Scar Ai <38693320+scarai-sap@users.noreply.github.com>
Co-authored-by: steinsebastian <54309425+steinsebastian@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: anjana-bl <anjana.b.l@sap.com>

leowangchao committed:
feat: Update Code Owners for /feature-libs/quote/  (#18869)

suprishi committed:
Feature/cxint 2989 download proposal (#18837)

Co-authored-by: Bose <saptarshi.bose01@sap.com>
Co-authored-by: anjana-bl <anjana.b.l@sap.com>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: Larisa Staroverova <larisa.staroverova@sap.com>

RadhepS committed:
fix: reduce cognitive complexity of function to resolve sonar warning (#18838)

fix: reduce cognitive complexity of function to resolve sonar warning

closes: https://jira.tools.sap/browse/CXSPA-7199

RadhepS committed:
chore: fix issue where Pull Requests could have been merged without E2Es passing (#18878)

chore: fix issue where Pull Requests could have been merged without E2Es passing

closes: https://jira.tools.sap/browse/CXSPA-7320

renovate[bot] committed:
chore(deps): update dependency webpack-cli to v5 (#18827)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency normalize-newline to v4 (#18868)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency commander to v12 (#18867)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Giancarlo Cordero Ortiz <46171897+giancorderoortiz@users.noreply.github.com>

i333055 committed:
CXSPA-7261: Accessibility issues for OTP login (#18876)

tiwariakshay55 committed:
CXSPA-6816: Fix schematics. (#18882)

kpawelczak committed:
chore: update dependency jest-preset-angular to v13.1.6  (#18829)

Co-authored-by: Radhep Sabapathipillai <34665674+RadhepS@users.noreply.github.com>

kimhw0630 committed:
chore: update webApplicationInjector.js (#18888)

Co-authored-by: Hakwoo Kim <hakwoo.kim@sap.com>

LindaSAP committed:
CXSPA-7325: Add scenarios which could not run on pipeline set into WTF D3 run (#18880)

Co-authored-by: Grace Dong <127385561+i53577@users.noreply.github.com>

scarai-sap committed:
Merge branch 'develop' into feature/CXSPA-3657

github-actions[bot] committed:
Add license header

scarai-sap committed:
Merge branch 'develop' of https://github.com/SAP/spartacus into feature/CXSPA-3657

scarai-sap committed:
Merge branch 'feature/CXSPA-3657' of https://github.com/SAP/spartacus into feature/CXSPA-3657

LindaSAP committed:
change backbutton element (#18891)

scarai-sap committed:
Merge branch 'develop' into feature/CXSPA-3657

eddyouedraogo committed:
CXEC-39725 : Enable guest checkout for all -spa-standalone stores ootb (#18870)

Pio-Bar committed:
fix: (CXSPA-1039) - Quick order search box role changed to combobox (#18859)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

Pio-Bar committed:
fix: (CXSPA-1062) - Comprehensive readout for slide indicators (#18842)

Co-authored-by: Piotr Bartkowiak <pbartkowiak@divante.pl>

giancorderoortiz committed:
Enable CodeQL.  (#18893)

kpawelczak committed:
Release 2211.24.0 1 (#18898)

Co-authored-by: Akshay Tiwari <tiwariakshay0@gmail.com>

scarai-sap committed:
Merge branch 'develop' into feature/CXSPA-3657
